src/coreComponents/functions/MultivariableTableFunction.hpp

```diff
@@ -0,0 +1,221 @@
+   */
+  arrayView1d< real64 const > getAxisMaximums() const { return m_axisMaximums.toViewConst(); }
+
+  /**
+   * @brief Get the table axes discretization points numbers
+   * @return a reference to an array of table axes discretization points numbers
+   */
+  arrayView1d< integer const > getAxisPoints() const { return m_axisPoints.toViewConst(); }
+
+
+  /**
+   * @brief Get the table axes step sizes
+   * @return a reference to an array of table axes step sizes
+   */
+  arrayView1d< real64 const > getAxisSteps() const { return m_axisSteps.toViewConst(); }
+
+  /**
+   * @brief Get the table axes step sizes inversions
+   * @return a reference to an array of table axes step sizes inversions
+   */
+  arrayView1d< real64 const > getAxisStepInvs() const { return m_axisStepInvs.toViewConst(); }
+
+  /**
+   * @brief Get the table axes hypercube index multiplicators
+   * @return a reference to an array of table axes hypercube index multiplicators
+   */
+  arrayView1d< globalIndex const > getAxisHypercubeMults() const { return m_axisHypercubeMults.toViewConst(); }
+
+  /**
+   * @brief Get the table values stored per-hypercube
+   * @return a reference to an array of table values stored per-hypercube
+   */
+  arrayView1d< real64 const > getHypercubeData() const { return m_hypercubeData.toViewConst(); }
+
+private:
+
+  /**
+   * @brief Get indexes of all vertices of a hypercube
+   *
+   * @param[in] hypercubeIndex index of a hypercube
+   * @param[out] hypercubePoints array of indexes of hypercube vertexes
+   */
+  void getHypercubePoints( globalIndex const hypercubeIndex, globalIndex_array & hypercubePoints ) const;
+
+  /// Number of table dimensions (inputs)
+  integer m_numDims;
+
+  /// Number of operators (functions to interpolate - ouputs)
+  integer m_numOps;
+
+  /// Number of vertices in each hypercube
+  integer m_numVerts;
+
+  /// Array [numDims] of axis minimum values
+  real64_array m_axisMinimums;
+
+  /// Array [numDims] of axis maximum values
+  real64_array m_axisMaximums;
+
+  /// Array [numDims] of axis discretization points numbers
+  integer_array m_axisPoints;
+
+  // inputs : data derived from table coordinates data
+
+  ///  Array [numDims] of axis interval lengths (axes are discretized uniformly)
+  real64_array m_axisSteps;
+
+  ///  Array [numDims] of inversions of axis interval lengths (axes are discretized uniformly)
+  real64_array m_axisStepInvs;
+
+  ///  Array [numDims] of point index mult factors for each axis
+  globalIndex_array m_axisPointMults;
+
+  ///  Array [numDims] of hypercube index mult factors for each axis
+  globalIndex_array m_axisHypercubeMults;
+
+  // inputs: operator sample data
+
+  ///  Main table data stored per point: all operator values for each table coordinate
+  real64_array m_pointData;
+
+  ///  Main table data stored per hypercube: all values required for interpolation withing give hypercube are stored contiguously
+  real64_array m_hypercubeData;
```

Comment:
@klevzoff, you are absolutely right regarding storage size requirements for this implementation and it definitely deserves consideration. Thank you for initiating this conversation! I`ve never explicitly tested this, but intuitively single "random access" to hypercube storage seemed faster than 2 ^d "random accesses" to point storage from the performance point of view. Saying "random" I mean that the index here will only depend on the state values of the corresponding mesh element (and not its geometrical location, for example). I don`t know the actual impact on performance, however, but it seems to me that it might be considerable for both CPU (because of cache) and GPU ( because of reading coalescence[which won`t occur in this implementation, but will occur in the next one, when single GPU thread is assigned per operator] and also cache). Regarding actual storage, it rarely exceeds tens of megabytes for most practical applications. However, since we only have static storage here, we can hit this limit much faster. 
I would prefer to test both point-hypercube storage use and multidimensional storage option with various layouts, but in a separate PR addressing optimization of interpolation performance / memory layout / footprint, while moving on for now with the current implementation. Please let me know if you find this acceptable...

Reply:
Yes, your plan is good, and the PR is in great shape to be merged IMO, so let's proceed. I definitely agree with you about cost of access (somewhat difficult for me to reason about GPU coalescing at the moment, but I totally trust your judgement on this, and also think the next implementation you're referring to will make things clearer for me).